docs/dev/documentation_guidelines.md

```diff
@@ -0,0 +1,158 @@
+     ```python
+     """
+     An example to achieve link effect in summary for :class:`~promptflow.xxx.MyClass`
+     For function, use :meth:`~promptflow.xxx.my_func`
+     """
+     ```
+
+7. There are some tricks to highlight the content in your docstring:
+    - Single backticks (`): Single backticks are used to represent inline code elements within the text. It is typically used to highlight function names, variable names, or any other code elements within the documentation.
+    - Double backticks(``): Double backticks are typically used to highlight a literal value.
+
+
+### Write function docstring
+
+```python
+from typing import Optional
+
+def my_method(param_int: Optional[int] = None) -> int:
+    """One-line summary
+
+    Detailed explanations.
+
+    :param param_int: Description of the parameter.
+    :type param_int: int
+    :raises [ErrorType1]: [ErrorDescription1]
+    :raises [ErrorType2]: [ErrorDescription2]
+    :return: Description of the return value.
+    :rtype: int
+    """
+    ...
+```
+
+In addition to `class docstring` notes:
+
+1. Function docstring should include return values.
+    - If return type is promptflow class, we should also use `~promptflow.xxx.[ClassName]`.
+2. Function docstring should include exceptions that may be raised in this function.
+    - If exception type is `PromptflowException`, use `~promptflow.xxx.[ExceptionName]`
+    - If multiple exceptions are raised, just add new lines of `:raises`, see the example above.
+
+
+## How to build doc site locally
+
+You can build the documentation site locally to preview the final effect of your docstring on the rendered site. This will provide you with a clear understanding of how your docstring will appear on our site once your changes are merged into the main branch.
+
+1. Setup your dev environment, see [dev_setup](./dev_setup.md) for details. Sphinx will load all source code to process docstring.
+2. Install `langchain` package since it is used in our code but not covered in `dev_setup`.
+3. Open a `powershell`, activate the conda env and navigate to `<repo-root>/scripts/docs` , run `doc_generation.ps1`:
+    ```pwsh
+    cd scripts\docs
+    .\doc_generation.ps1 -WithReferenceDoc -WarningAsError
+    ```
+    - For the first time you execute this command, it will take some time to install `sphinx` dependencies. After the initial installation, next time you can add param `-SkipInstall` to above command to save some time for dependency check.
+4. Check warnings/errors in the build log, fix them if any, then build again.
+5. Open `scripts/docs/_build/index.html` to preview the local doc site.
+
+## Aditional comments
+
+- **Utilities**: The [autoDocstring](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring) VSCode extension or GitHub Copilot can help autocomplete in this style for you.
```

Comment:
The most practical user guide is here. can we put this at the very head?

Reply:
you mean autoDocstring link? how could that be the most practical user guide?

Reply:
Offline synced, will keep the way it is.